libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,748 @@
+#include <arrow/python/numpy_to_arrow.h>
+#include <arrow/python/python_to_arrow.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "katana/python/Conventions.h"
+#include "katana/python/CythonIntegration.h"
+#include "katana/python/EntityTypeManagerPython.h"
+#include "katana/python/ErrorHandling.h"
+#include "katana/python/NumbaSupport.h"
+#include "katana/python/PropertyGraphPython.h"
+#include "katana/python/PythonModuleInitializers.h"
+#include "katana/python/TemplateSupport.h"
+#include "katana/python/TypeBans.h"
+
+namespace py = pybind11;
+
+katana::Result<std::shared_ptr<arrow::Table>>
+katana::python::PythonArgumentsToTable(
+    const pybind11::object& table, const pybind11::dict& kwargs) {
+  std::shared_ptr<arrow::Table> arrow_table;
+
+  if (arrow::py::is_table(table.ptr())) {
+    arrow_table = KATANA_CHECKED(arrow::py::unwrap_table(table.ptr()));
+  } else if (!table.is_none()) {
+    arrow_table = KATANA_CHECKED(arrow::py::unwrap_table(
+        py::module::import("pyarrow").attr("table")(table).ptr()));
+  }
+
+  for (const auto& [k, v] : kwargs) {
+    std::shared_ptr<arrow::ChunkedArray> array;
+    if (py::isinstance<py::array>(v)) {
+      // Convert a numpy array.
+      py::array pyarray = py::cast<py::array>(v);
+      std::shared_ptr<arrow::DataType> dtype;
+      KATANA_CHECKED(
+          arrow::py::NumPyDtypeToArrow(pyarray.dtype().ptr(), &dtype));
+      KATANA_CHECKED(arrow::py::NdarrayToArrow(
+          arrow::default_memory_pool(), v.ptr(), nullptr, true, dtype,
+          arrow::compute::CastOptions::Safe(dtype), &array));
+      KATANA_LOG_DEBUG_ASSERT(array);
+    } else {
+      // Convert any other python sequence.
+      arrow::py::PyConversionOptions options;
+      options.from_pandas = true;
+      array = KATANA_CHECKED(
+          arrow::py::ConvertPySequence(v.ptr(), nullptr, options));
+    }
+    auto field = arrow::field(py::cast<py::str>(k), array->type());
+    if (arrow_table) {
+      arrow_table = KATANA_CHECKED(
+          arrow_table->AddColumn(arrow_table->num_columns(), field, array));
+    } else {
+      arrow_table = arrow::Table::Make(arrow::schema({field}), {array});
+    }
+  }
+
+  if (!arrow_table) {
+    return KATANA_ERROR(
+        katana::ErrorCode::InvalidArgument,
+        "A table argument (a dict or arrow table) or a keyword argument must "
+        "be provided.");
+  }
+
+  return arrow_table;
+}
+
+class LazyDataAccessor {
+public:
+  virtual ~LazyDataAccessor();
+  virtual py::object at(ssize_t i) const = 0;
+  virtual py::object array(const py::slice& slice) const = 0;
+};
+
+LazyDataAccessor::~LazyDataAccessor() = default;
+
+template <typename T>
+class LazyDataAccessorTyped : public LazyDataAccessor {
+public:
+  virtual T at_typed(ssize_t i) const = 0;
+  virtual py::object at(ssize_t i) const final { return py::cast(at_typed(i)); }
+  virtual py::object array(const py::slice& slice) const {
```

Comment:
Nitpick: you don't need `virtual` is derived classes, just `override` or `final`.

Reply:
Aaagh. I tried to get this right and failed anyway. I'll fix this. :-) Thanks.